config/config.go

```diff
@@ -263,6 +258,111 @@ func FromViper(v *viper.Viper) (*Config, error) {
 	return cfg, nil
 }
 
+func determineTreeRoot(v *viper.Viper, cfg *Config) error {
+	var err error
+
+	// enforce the various tree root options are mutually exclusive
+	// some of this is being done for us at the flag level, but you can also set these values in config or environment
+	// variables.
+	count := 0
+
+	if cfg.TreeRoot != "" {
+		count++
+	}
+
+	if cfg.TreeRootCmd != "" {
+		count++
+	}
+
+	if cfg.TreeRootFile != "" {
+		count++
+	}
+
+	if count > 1 {
+		return errors.New("only one of tree-root, tree-root-cmd or tree-root-file can be specified")
+	}
+
+	// set git-based tree root command if the walker is git and no tree root has been specified
+	if cfg.Walk == walk.Git.String() && count == 0 {
+		cfg.TreeRootCmd = "git rev-parse --show-toplevel"
+
+		log.Infof(
+			"git walker enabled and tree root has not been specified: defaulting tree-root-cmd to '%s'",
+			cfg.TreeRootCmd,
+		)
+	}
+
+	switch {
+	case cfg.TreeRoot != "":
+		log.Debugf("tree root specified explicitly: %s", cfg.TreeRoot)
+
+	case cfg.TreeRootFile != "":
+		log.Debugf("searching for tree root using --tree-root-file: %s", cfg.TreeRootFile)
+
+		_, cfg.TreeRoot, err = FindUp(cfg.WorkingDirectory, cfg.TreeRootFile)
+		if err != nil {
+			return fmt.Errorf("failed to find tree-root based on tree-root-file: %w", err)
+		}
+
+	case cfg.TreeRootCmd != "":
+		log.Debugf("searching for tree root using --tree-root-cmd: %s", cfg.TreeRootCmd)
+
+		if cfg.TreeRoot, err = execTreeRootCmd(cfg); err != nil {
+			return err
+		}
+
+	default:
+		// no tree root was specified
+		log.Debugf("no tree root specified, defaulting to the directory containing the config file: %s", v.ConfigFileUsed())
+
+		cfg.TreeRoot = filepath.Dir(v.ConfigFileUsed())
+	}
+
+	// resolve tree root to an absolute path
+	if cfg.TreeRoot, err = filepath.Abs(cfg.TreeRoot); err != nil {
+		return fmt.Errorf("failed to get absolute path for tree root: %w", err)
+	}
+
+	log.Debugf("tree root: %s", cfg.TreeRoot)
+
+	return nil
+}
+
+func execTreeRootCmd(cfg *Config) (string, error) {
+	// split the command first, resolving any '' and "" entries
+	parts, splitErr := shlex.Split(cfg.TreeRootCmd)
+	if splitErr != nil {
+		return "", fmt.Errorf("failed to parse tree-root-cmd: %w", splitErr)
+	}
+
+	// set a reasonable timeout of 2 seconds to wait for the command to return
+	// it shouldn't take anywhere near this amount of time unless there's a problem
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+	defer cancel()
+
+	// construct the command, setting the correct working directory
+	//nolint:gosec
+	cmd := exec.CommandContext(ctx, parts[0], parts[1:]...)
+	cmd.Dir = cfg.WorkingDirectory
+
+	// execute
+	out, cmdErr := cmd.CombinedOutput()
+	if cmdErr != nil {
+		log.Errorf("tree-root-cmd output: \n%s", out)
+
+		return "", fmt.Errorf("failed to run tree-root-cmd: %w", cmdErr)
+	}
+
+	// trim the output and check it's not empty
```

Comment:
Should we also check the output is not multi-line? i.e. has no `"\n"` infix.

And/or check that the output is a real directory - or I guess that is handled later?

Reply:
Multiline output will force the resolution to an absolute path to bork later on. That being said, it wouldn't hurt to catch this specifically and provide pointed feedback that the script or command being used is not behaving as expected.

Reply:
Added a test for this and an explicit error